torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -0,0 +1,114 @@
+from functools import reduce
+import torch
+from torch.fx.tensor_type import Dyn, is_consistent, TensorType, is_more_precise
+import operator
+
+
+class GraphTypeChecker:
+    def __init__(self, env, traced):
+        self.env = env
+        self.traced = traced
+
+    def type_check(self):
+        """
+        A gradual type checker for graphs
+        Effect: every node's field type will be
+        populated with a type after type-checking is done
+        """
+        # here we collect types for variables
+        graph = self.traced.graph
+
+        # type check every node with gradual type rules
+        # if any node does not type check return false
+        for n in graph.nodes:
+            if not self.type_check_node(n):
+                return False
+        return True
+
+    def type_check_node(self, n):
+        """
+        Type check a given node.
+        Current operations:
+        - Reshape
+        - Transpose
+        - Add
+        """
+        if n.op == 'placeholder':
+            if n.type is None:
+                n.type = Dyn
+            else:
+                return n.type
+
+        if n.op == 'call_function':
+            if n.target == torch.add or n.target == operator.add:
```

Comment:
To make this a bit cleaner, we could define a system where we can register these rules as a separate function. For example:

```
import torch, operator
import torch.fx
from typing import Callable, Dict

_INFERENCE_RULES : Dict[torch.fx.node.Target, Callable] = {}

def register_inference_rule(call_target):
    def register(fn):
        if call_target in _INFERENCE_RULES:
            raise RuntimeError('Inference rule already registered for {call_target}!')
        _INFERENCE_RULES[call_target] = fn
        return fn
    return register


@register_inference_rule(torch.add)
@register_inference_rule(operator.add)
def add_inference_rule(n : torch.fx.Node):
    # Type inference rule for add here


@register_inference_rule(torch.reshape)
def reshape_inference_rule(n : torch.fx.Node):
    # Type inference for reshape here
```

Then here we can replace the `if/else` tree with something like:

```
if n.target in _INFERENCE_REGISTRY:
  return _INFERENCE_REGISTRY[n.target](n)
else:
  raise RuntimeError(f'No inference rule registered for target {n.target}!')
```